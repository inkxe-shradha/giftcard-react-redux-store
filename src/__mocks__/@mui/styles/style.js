import * as Styles from "@mui/styles";

const makeStyles = () => {
  return () => {
    /**
     * Note: if you want to mock this return value to be
     * different within a test suite then use
     * the pattern defined here:
     * https://jestjs.io/docs/en/manual-mocks
     **/
    return {};
  };
};

module.exports = { ...Styles, makeStyles };
