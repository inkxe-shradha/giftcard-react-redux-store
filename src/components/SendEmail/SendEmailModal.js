import React from "react";
import Modal from "@mui/material/Modal";
import Container from "@mui/material/Container";
import PropTypes from "prop-types";
import Box from "@mui/material/Box";
import { makeStyles } from "@mui/styles";
import IconButton from "@mui/material/IconButton";
import CloseIcon from "@mui/icons-material/Close";
import SendIcon from "@mui/icons-material/Send";
import Typography from "@mui/material/Typography";
import { TextField } from "@mui/material";
import { useFormik } from "formik";
import { LoadingButton } from "@mui/lab";
import Alert from "@mui/material/Alert";
import * as yup from "yup";

const style = {
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  bgcolor: "background.paper",
  border: "2px solid #000",
  boxShadow: 24,
  p: 4,
};

const useStyles = makeStyles((theme) => ({
  modalBox: {
    [theme.breakpoints.down("md")]: {
      width: "100%",
    },
    [theme.breakpoints.up("md")]: {
      width: "60%",
    },
    [theme.breakpoints.up("lg")]: {
      width: "60%",
    },
  },
  iconButton: {
    textAlign: "right",
    float: "right",
  },
}));
const validationSchema = yup.object({
  email: yup.string().email("Invalid email").required("Email is required"),
});

const SendEmailModal = ({
  isOpen,
  handleClose,
  handleSubmit,
  errorMessage,
  loading,
}) => {
  const initialValues = {
    email: "",
  };
  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit: (values) => {
      handleSubmit(values);
    },
  });
  const classes = useStyles();
  return (
    <Modal
      open={isOpen}
      onClose={handleClose}
      aria-labelledby="Modal for opening the Sending the modal"
      aria-describedby="short modal form descriptions"
    >
      <Container maxWidth="sm">
        <Box sx={{ ...style }} className={classes.modalBox}>
          <IconButton
            className={classes.iconButton}
            aria-label="close"
            onClick={handleClose}
          >
            <CloseIcon />
          </IconButton>
          <Typography id="modal-modal-title" variant="h6" component="h2">
            Send Email
          </Typography>
          {errorMessage && (
            <Alert severity="error" className="my-2">
              {errorMessage}
            </Alert>
          )}
          <form onSubmit={formik.handleSubmit} noValidate>
            <div className="row">
              <div className="col-md-8 offset-md-2">
                <TextField
                  fullWidth
                  id="email"
                  name="email"
                  label="Sender Email"
                  value={formik.values.email}
                  onChange={formik.handleChange}
                  error={formik.touched.email && Boolean(formik.errors.email)}
                  helperText={formik.touched.email && formik.errors.email}
                  required
                  margin="normal"
                />
              </div>
            </div>

            <div className="row mt-2">
              <div className="col-md-12 text-center">
                <LoadingButton
                  id="sendEmailBtn"
                  loading={loading}
                  type="submit"
                  variant="contained"
                  margin="normal"
                  sx={{ margin: 2 }}
                  endIcon={<SendIcon />}
                >
                  Send
                </LoadingButton>
              </div>
            </div>
          </form>
        </Box>
      </Container>
    </Modal>
  );
};

// PropTypes validation
SendEmailModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string,
};

export default SendEmailModal;
