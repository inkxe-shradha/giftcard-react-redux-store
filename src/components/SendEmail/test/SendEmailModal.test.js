import { createTheme } from "@mui/material";
import { ThemeProvider } from "@mui/styles";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { mount } from "enzyme";
import SendEmailModal from "../SendEmailModal";
describe("Test suite for the Email Sending Modal", () => {
  const defaultProps = {
    isOpen: true,
    handleClose: jest.fn(),
    handleSubmit: jest.fn(),
    errorMessage: "",
    loading: false,
  };
  const theme = createTheme();
  const customRender = (type = "render", props = defaultProps) => {
    if (type === "render") {
      return render(
        <ThemeProvider theme={theme}>
          <SendEmailModal isOpen={true} {...props} />
        </ThemeProvider>
      );
    } else {
      return mount(
        <ThemeProvider theme={theme}>
          <SendEmailModal {...props} />
        </ThemeProvider>
      );
    }
  };

  it("Should create a snapshot of the SendEmailModal", () => {
    const { asFragment } = customRender();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should contain a form and a input field of type email", () => {
    const view = customRender("mount");
    expect(view.find("form").length).toBe(1);
    expect(view.find("input[name='email']").length).toBe(1);
  });

  it("Should have show Invalid email if the email address is no valid", async () => {
    customRender();
    const emailElement = screen.getByRole("textbox", { type: "text" });
    fireEvent.change(emailElement, { target: { value: "test" } });
    // Button click
    const button = screen.getByRole("button", { name: "Send" });
    fireEvent.click(button);
    await waitFor(() => {
      const emailErrorAfter = screen.getByText("Invalid email");
      expect(emailErrorAfter).toBeInTheDocument();
    });
  });

  it("Should be valid if a valid email address sent", async () => {
    customRender();
    const emailElement = screen.getByRole("textbox", { type: "text" });
    fireEvent.change(emailElement, { target: { value: "dummy@gmail.com" } });
    // Button click
    const button = screen.getByRole("button", { name: "Send" });
    fireEvent.click(button);
  });
});
