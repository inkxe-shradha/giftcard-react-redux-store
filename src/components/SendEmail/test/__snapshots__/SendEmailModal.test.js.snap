// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for the Email Sending Modal Should create a snapshot of the SendEmailModal 1`] = `<DocumentFragment />`;
